src/types.h

```diff
@@ -261,40 +261,38 @@ enum Rank : int {
 /// care to avoid left-shifting a signed int to avoid undefined behavior.
 enum Score : int { SCORE_ZERO };
 
-inline Score make_score(int mg, int eg) {
+inline constexpr Score make_score(int mg, int eg) {
   return Score((int)((unsigned int)eg << 16) + mg);
 }
 
 /// Extracting the signed lower and upper 16 bits is not so trivial because
 /// according to the standard a simple cast to short is implementation defined
-/// and so is a right shift of a signed integer.
-inline Value eg_value(Score s) {
-
-  union { uint16_t u; int16_t s; } eg = { uint16_t(unsigned(s + 0x8000) >> 16) };
-  return Value(eg.s);
+/// and so is a right shift of a signed integer. Note: a cast to a short is
+// only implementation-defined if the source value can't be represented in the
+// destination type.
+inline constexpr Value eg_value(Score s) {
+	return Value(int16_t(unsigned(s + 0x8000) >> 16));
 }
 
-inline Value mg_value(Score s) {
-
-  union { uint16_t u; int16_t s; } mg = { uint16_t(unsigned(s)) };
-  return Value(mg.s);
+inline constexpr Value mg_value(Score s) {
+	return Value(int16_t(unsigned(s & 0xffff)));
```

Comment:
This is adding an unnecessary AND instruction compared to master and is executed extremely frequently.

Reply:
return Value(int16_t(uint16_t(unsigned(s)))); is equivalent w/o the AND.

Reply:
You're right, but if the score from which the mg value has to be extracted is a constant, then the mg_value() function takes no shifts or and-operations at run-time at all, they're optimized all the way down into direct loads of constants.

Reply:
Do we need to touch the internal implementation of mg_value() at all for the constexpr patch?

Reply:
In the c++11 standard, constexpr functions can only consist of return statements, so you can't do operations on unions in constexpr functions (in c++14 you can do a lot more). I could easily revert Value mg_value(Score s) to the old version if you think the constexpr version is ugly or creates unnecessary overhead,, but on my hardware at least, it seems a fraction faster.

Reply:
I think the most recent discussion on eg_value() and mg_value() is here: https://github.com/official-stockfish/Stockfish/pull/849

If I can believe what I wrote there, the current union-based implementation would be legal in C11 but relies on a gcc extension for C++11.

That was not found to be a good argument then (also in view of many earlier attempts and discussions). But this pull request gives a new argument for changing their implementation, namely to make them constexpr.

Let's check mg_value() in godbolt: https://godbolt.org/g/F5shzs

It translates to a single movswl, which seems optimal.
Btw, this is not because of the constexpr keyword (remove it and see for yourself). I am still quite confident that the constexpr keyword in Stockfish makes a difference only for the initialisation of static constants. Since there is no place in Stockfish where a static constant is initialised by taking the mg_value() of a constant Score value, making mg_value() constexpr does not actually change anything (but I am ready to be proven wrong on this by a counterexample!). Still, it is probably not a bad idea to use constexpr where it potentially makes sense, such as here.

Reply:
Making mg_value() constexpr allows functions that use it to be constexpr too, in this case that's just the Score operator/(Score s, int i), so that's why i don't object to reverting much. Reverting make_score() would be a much bigger issue for me, because the constexpr version has clear benefits over the non-constexpr version.

Reply:
@gvreuls Not all calls to mg_value() can be evaluated at compile time.  That is why I object to the unnecessary AND (not the constexpr).  You can rewrite just as easily w/o it as I showed.

Reply:
@mstembera If I simply leave the union operations intact and change the inline specifier into a constexpr specifier I ger compiler errors telling me the functions are not valid constexpr functions because they contain more than just a return statement:

types.h: In function ‘constexpr Value eg_value(Score)’:
types.h:285:1: error: body of constexpr function ‘constexpr Value eg_value(Score)’ not a return-statement

types.h: In function ‘constexpr Value mg_value(Score)’:
types.h:290:1: error: body of constexpr function ‘constexpr Value mg_value(Score)’ not a return-statement

So if it's going to the old union operations instead of bit operations, we have to drop the constexpr and keep the functions inlined.

Reply:
What's wrong with just `return Value(int16_t(uint16_t(s)));`? It compiles to the same assembly code as master: https://godbolt.org/g/HkoZNd